src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -125,114 +126,248 @@ EmbeddedAssemblies::TypeMappingInfo_compare_key (const void *a, const void *b)
+int
+EmbeddedAssemblies::compare_java_name (const char *java_name, const TypeMapJava *entry)
 {
-	TypeMappingInfo *p        = new TypeMappingInfo (); // calloc (1, sizeof (struct TypeMappingInfo));
-	int              version  = 0;
-	const char      *data     = addr;
-
-	extract_int (&data, source_apk, source_entry, "version",   &version);
-	if (version != 1) {
-		delete p;
-		log_warn (LOG_DEFAULT, "Unsupported version '%i' within type mapping file '%s!%s'. Ignoring...", version, source_apk, source_entry);
-		return false;
+	return strcmp (java_name, reinterpret_cast<const char*>(entry->java_name));
+}
+
+#if defined (DEBUG) || !defined (ANDROID)
+int
+EmbeddedAssemblies::compare_java_name (const char *java_name, const uint8_t *entry)
+{
+	return strcmp (java_name, reinterpret_cast<const char*>(entry));
+}
+#endif // DEBUG || !ANDROID
+
+const char*
+EmbeddedAssemblies::typemap_managed_to_java (const uint8_t *mvid, const int32_t token)
+{
+	timing_period total_time;
+	if (XA_UNLIKELY (utils.should_log (LOG_TIMING))) {
+		timing = new Timing ();
+		total_time.mark_start ();
 	}
 
-	extract_int (&data, source_apk, source_entry, "entry-count",  &p->num_entries);
-	extract_int (&data, source_apk, source_entry, "entry-len",    &p->entry_length);
-	extract_int (&data, source_apk, source_entry, "value-offset", &p->value_offset);
-	p->mapping      = data;
+	if (mvid == nullptr) {
+		return nullptr;
+	}
 
-	if ((p->mapping == 0) ||
-			(p->num_entries <= 0) ||
-			(p->entry_length <= 0) ||
-			(p->value_offset >= p->entry_length) ||
-			(p->mapping == nullptr)) {
-		log_warn (LOG_DEFAULT, "Could not read type mapping file '%s!%s'. Ignoring...", source_apk, source_entry);
-		delete p;
-		return false;
+	const TypeMapModule *map;
+	uint32_t map_entry_count;
+#if defined (DEBUG) || !defined (ANDROID)
+	if (application_config.instant_run_enabled) {
+		map = modules;
+		map_entry_count = module_count;
+	} else {
+#endif
+		map = map_modules;
+		map_entry_count = map_module_count;
+#if defined (DEBUG) || !defined (ANDROID)
+	}
+#endif
+	const TypeMapModule *match = binary_search<uint8_t, TypeMapModule, compare_mvid> (mvid, map, map_entry_count);
+	if (match == nullptr) {
+		log_warn (LOG_ASSEMBLY, "typemap: module matching MVID [%s] not found.", mono_guid_to_string (mvid));
+		return nullptr;
 	}
 
-	p->source_apk   = strdup (source_apk);
-	p->source_entry = strdup (source_entry);
-	if (*info) {
-		(*info)->next = p;
+	if (match->map == nullptr) {
+		log_warn (LOG_ASSEMBLY, "typemap: module with MVID [%s] has no associated type map.", mono_guid_to_string (mvid));
+		return nullptr;
+	}
+
+	// Each map entry is a pair of 32-bit integers: [TypeTokenID][JavaMapArrayIndex]
+	const TypeMapModuleEntry *entry = binary_search <int32_t, TypeMapModuleEntry, compare_type_token> (&token, match->map, match->entry_count);
+	if (entry == nullptr) {
+		if (match->duplicate_count > 0 && match->duplicate_map != nullptr) {
```

Comment:
I don't understand why there's a "duplicate map" in the first place...

Reply:
Because we don't use type names anymore, we use tokens - there might be multiple managed type tokens referring to a single Java type where previously they would all use the same **name**